### ISSUE TITLE
MAster-Deatil Form, Need Some Help.

### ISSUE THREAD

**Issue by mlutzow** [volkanceylan/Serenity#906]
've a Master Detail Editor that seems to work, partially, I can't get the detail form to populate however.

This is what I've done with the code, to date. Hopefully I've missed something very simple, if I need to write some code to get the records, that's fine as well, but doesn't seem to be required based on the tutorials, etc., etc.

Each table has a field defined as int32, identity, primary key, auto increment (named BudgetKey)
The Detail table has an additional field defined as int32, foreign key field. This field contains the primary key value of the master table of the relationship.
I can define the relationship via an alternate method but the relationship is a bit more complex.

Master Part of the Relation:

RevenueSummaryRow.cs:
//
....
....
       [DisplayName("Distribution Type"), Column("Distribution_Type"), Size(2)]
        public String DistributionType
        {
            get { return Fields.DistributionType[this]; }
            set { Fields.DistributionType[this] = value; }
        }

        [DisplayName("Version")]
        public Int32? Version
        {
            get { return Fields.Version[this]; }
            set { Fields.Version[this] = value; }
        }
        [DisplayName("Periods"), MasterDetailRelation(foreignKey: "RevenueSummaryLink"), ClientSide]
        public List<RevenuePeriodsRow> PeriodList
        {
            get { return Fields.PeriodList[this]; }
            set { Fields.PeriodList[this] = value; }
        }
...
...
            public RowListField<RevenuePeriodsRow> PeriodList;
            public RowFields()
                : base("[budgets].[Revenue_Summary]")
...
//
RevenueSummaryFoirm.cs:
//
...
        public Int32 Version { get; set; }

        [Category("Periods")]
        [RevenuePeriodsEditor]
        public List<Entities.RevenuePeriodsRow> RevenuePeriodList { get; set; }

        #region HiddenFields
...
//
The Detail Portion of the Relationship
RevenuePeriodsRow.cs:
//
...
        //[Hidden]
        [DisplayName("Budget Key"), PrimaryKey, Identity]
        public Int32? BudgetKey
        {
            get { return Fields.BudgetKey[this]; }
            set { Fields.BudgetKey[this] = value; }
        }

        //[Hidden]
        [DisplayName("Revenue Summary Link"), NotNull, ForeignKey("[budgets].Revenue_Summary", "BudgetKey")]
        [LeftJoin("jRevenueSummaryLink"), TextualField("RevenueSummaryLinkCustomerNumber")]
        public Int32? RevenueSummaryLink
        {
            get { return Fields.RevenueSummaryLink[this]; }
            set { Fields.RevenueSummaryLink[this] = value; }
        }
I created the following
RevenuePeriodEditor.ts:

/// <reference path="../../../Common/Helpers/GridEditorBase.ts" />

namespace HgiBudgets.HGIBudgetsDB {

	@Serenity.Decorators.registerClass()
	export class RevenuePeriodsEditor extends Common.GridEditorBase<RevenuePeriodsRow> {
		protected getColumnsKey() { return "HGIBudgetsDB.Revenue.RevenuePeriods"; }

		protected getDialogType() { return RevenuePeriodsDialog; }

		protected getLocalTextPrefix() { return RevenuePeriodsRow.localTextPrefix; }

		constructor(container: JQuery) {
			super(container);
		}

		protected getAddButtonCaption() {
			return "Add Period";
		}
	}
}

Any assistance is greatly appreciated.

**Comment by volkanceylan** [comment]
Is detail grid empty or form? Do you have any errors in browser console?

**Comment by jsbUSMC** [comment]
If you have any expression fields in RevenuePeriodRow that need to be available in your RevenuePeriodEditor, you need to add them to the `IncludeColumns: "Field1, Field2, ...." property of your MasterDetailRelationship attribute.

**Comment by mlutzow** [comment]
To volkanceylan:

The form populates with the test data I'm using, both the Revenue Summary and the Revenue Periods.

The Detail grid portion of the form is empty, however.

**Comment by mlutzow** [comment]
to jsbUSMC:
 Scratch my last question, I've figured it out.

Thank you for your response. I'm in learning mode with the C# language and the SERENE Template, so I'll probably have a few more questions, along the way.

**Comment by mlutzow** [comment]
I've added the following code under the RetrieveRequestHandler:
"        private class MyRetrieveHandler : RetrieveRequestHandler<MyRow>
        {
            protected override void OnReturn()
            {
                base.OnReturn();
                var revper = Entities.RevenuePeriodsRow.Fields;
                Row.PeriodList = Connection.List<Entities.RevenuePeriodsRow>(q => q
                .SelectTableFields()
                .Where(revper.RevenueSummaryLink == Row.BudgetKey.Value));
            }
        }
The Detail Grid was still empty, I ran through using some debugging and found out that the system appears to be selecting the records, at least in the params section, the value was the correct value that I had selected for the form.  Further in the 'Autos' section of the debugger for the PeriodList the count was correct for the number of records that should have been selected and expanding the items under the PeriodList, the data returned is correct (I've tried attaching a small image to illustrate).

I'm stumped!

![revperdebugsession](https://cloud.githubusercontent.com/assets/19471794/16808856/d344dfda-48e3-11e6-9cb8-4238a99dd16d.PNG)
.

I've noticed a couple of errors, that are possibly of no consequence, the first is:
![staticmemberserror_20160713](https://cloud.githubusercontent.com/assets/19471794/16809020/560fe4be-48e4-11e6-8118-d95e52bc0421.PNG)

The next errors deal with fields that were created by sergen when I created the Revenue Periods form.
![fielderrors_20160713](https://cloud.githubusercontent.com/assets/19471794/16809108/aacf2672-48e4-11e6-8593-b78b7cefe8f4.PNG)

**Comment by jsbUSMC** [comment]
Shouldn't `Row.PeriodList` be `Row.RevenuePeriodList`?

**Comment by volkanceylan** [comment]
Put a Gist of your MasterRow, DetailRow, MasterForm, MasterDialog, you probably messed things too much

**Comment by mlutzow** [comment]
Gentlemen, I have found the issue, to many late nights doing C#, Entity Framework, and ASP.NET MVC 5 tutorials and then working on the program with the SERENE manual in hand.

I had defined the list as RevenuePeriodList in the Master Form and as PeriodList everywhere else. I changed the name to RevenuePeriodsList in all locations and now the Detail Grid is populating with the correct data.

Thank you both so much for your input and help to assist with finding the issue.

volkanceylan, I haven't posted the Gist because the issue has been resolved, but if you'd still like it I'd be more than happy to post the items you requested.
